executor/src/witgen/jit/witgen_inference.rs

```diff
@@ -89,38 +92,81 @@ impl<'a, T: FieldElement, FixedEval: FixedEvaluator<T>> WitgenInference<'a, T, F
+        // that value is stored in the range constraints.
+        let rc = self.range_constraint(variable.clone());
+        if let Some(val) = rc.as_ref().and_then(|rc| rc.try_to_single_value()) {
+            val.into()
+        } else if self.known_variables.contains(&variable) {
+            AffineSymbolicExpression::from_known_symbol(variable, rc)
+        } else {
+            AffineSymbolicExpression::from_unknown_variable(variable, rc)
+        }
+    }
+
+    /// Process the constraint that the expression evaluated at the given offset equals the given value.
+    /// This does not have to be solvable right away, but is always processed as soon as we have progress.
+    /// Note that either the expression or the value might contain unknown variables.
+    pub fn assign_constant(
         &mut self,
-        expression: &Expression<T>,
+        expression: &'a Expression<T>,
         offset: i32,
-        value: AffineSymbolicExpression<T, Variable>,
-    ) -> Result<(), String> {
-        let affine_expression = self
-            .evaluate(expression, offset)
-            .ok_or_else(|| format!("Expression is not affine: {expression}"))?;
-        let result = (affine_expression - value.clone())
-            .solve()
-            .map_err(|err| format!("Could not solve ({expression} - {value}): {err}"))?;
-        match self.ingest_effects(result).complete {
-            true => Ok(()),
-            false => Err("Wasn't able to complete the assignment".to_string()),
+        value: T,
+    ) -> ProcessSummary {
+        self.assign(expression, offset, VariableOrValue::Value(value))
+    }
+
+    /// Process the constraint that the expression evaluated at the given offset equals the given formal variable.
+    /// This does not have to be solvable right away, but is always processed as soon as we have progress.
+    /// Note that either the expression or the value might contain unknown variables.
+    pub fn assign_variable(
+        &mut self,
+        expression: &'a Expression<T>,
+        offset: i32,
+        variable: Variable,
+    ) -> ProcessSummary {
+        self.assign(expression, offset, VariableOrValue::Variable(variable))
+    }
+
+    pub fn has_unsolved_assignments(&self) -> bool {
+        !self.assignments.is_empty()
+    }
+
+    fn assign(
+        &mut self,
+        expression: &'a Expression<T>,
+        offset: i32,
+        var_or_val: VariableOrValue<T, Variable>,
+    ) -> ProcessSummary {
+        let v = match &var_or_val {
+            VariableOrValue::Variable(v) => self.variable_to_expression(v.clone()),
+            VariableOrValue::Value(v) => (*v).into(),
+        };
+        let r = self.process_polynomial_identity(expression, offset, v);
+        let summary = self.ingest_effects(r);
+        if !summary.complete {
+            // Put it into the queue.
+            self.assignments.push((expression, offset, var_or_val));
         }
+        summary
     }
 
     fn process_polynomial_identity(
```

Comment:
This name makes less sense now. I think if you externalize the assignments (one of my suggestions below), we can merge this function and `assign`, and call it `assign`.